driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,681 @@
+            final InternalConnection connection,
+            final ConnectionDescription connectionDescription,
+            final Function<byte[], byte[]> evaluateChallengeFunction) {
+        this.evaluateChallengeFunction = evaluateChallengeFunction;
+        super.authenticate(connection, connectionDescription);
+    }
+
+    private void authLock(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        Locks.withLock(getMongoCredentialWithCache().getOidcLock(), () -> {
+            while (true) {
+                try {
+                    authenticateUsing(connection, connectionDescription, (challenge) -> evaluate(challenge));
+                    break;
+                } catch (MongoSecurityException e) {
+                    boolean shouldRetry = triggersRetry(e) && shouldRetryHandler();
+                    if (!shouldRetry) {
+                        throw e;
+                    }
+                }
+            }
+            return null;
+        });
+    }
+
+    private byte[] evaluate(final byte[] challenge) {
+        if (isAutomaticAuthentication()) {
+            return prepareAwsTokenFromFile();
+        }
+
+        OidcRequestCallback requestCallback = assertNotNull(getRequestCallback());
+        MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+        OidcCacheEntry cacheEntry = mongoCredentialWithCache.getOidcCacheEntry();
+        String cachedAccessToken = getValidCachedAccessToken();
+        String invalidConnectionAccessToken = connectionLastAccessToken;
+        String cachedRefreshToken = cacheEntry.getRefreshToken();
+        IdpInfo cachedIdpInfo = cacheEntry.getIdpInfo();
+
+        if (cachedAccessToken != null) {
+            boolean cachedTokenIsInvalid = cachedAccessToken.equals(invalidConnectionAccessToken);
+            if (cachedTokenIsInvalid) {
+                mongoCredentialWithCache.setOidcCacheEntry(cacheEntry.clearAccessToken());
+                cachedAccessToken = null;
+            }
+        }
+        OidcRefreshCallback refreshCallback = getRefreshCallback();
+        if (cachedAccessToken != null) {
+            fallbackState = FallbackState.PHASE_1_CACHED_TOKEN;
+            return prepareTokenAsJwt(cachedAccessToken);
+        } else if (refreshCallback != null && cachedRefreshToken != null) {
+            assertTrue(cachedIdpInfo != null);
+            // Invoke Refresh Callback using cached Refresh Token
+            validateAllowedHosts(getMongoCredential());
+            fallbackState = FallbackState.PHASE_2_REFRESH_CALLBACK_TOKEN;
+            IdpResponse result = refreshCallback.onRefresh(new OidcRefreshContextImpl(
+                    cachedIdpInfo, cachedRefreshToken, CALLBACK_TIMEOUT));
+            return populateCacheWithCallbackResultAndPrepareJwt(cachedIdpInfo, result);
+        } else {
+            // cache is empty
+            boolean idpInfoNotPresent = challenge.length == 0;
```

Comment:
This `challenge.length == 0` is a place where an extensive explanation comment is needed for future readers. You explained this to me, and I thought I sorta understood, but I can't reproduce the explanation now. I only remember that speculative authentication (`wrapInSpeculative(prepareUsername(userName))`) is involved here. Could you please add such an explanatory comment?

Reply:
Thank you!